python_modules/dagster/dagster/core/execution/plan/plan.py

```diff
@@ -758,13 +758,36 @@ def build_subset_plan(
         step_output_versions = check.opt_dict_param(
             step_output_versions, "step_output_versions", key_type=StepOutputHandle, value_type=str
         )
-        step_handles_to_execute = [StepHandle.parse_from_key(key) for key in step_keys_to_execute]
 
+        step_handles_to_execute_set: Set[StepHandleUnion] = set()
         bad_keys = []
-        for handle in step_handles_to_execute:
+        for key in step_keys_to_execute:
+            handle = StepHandle.parse_from_key(key)
+
             if handle not in self.step_dict:
+                if isinstance(handle, ResolvedFromDynamicStepHandle):
+                    unresolved_handle = handle.unresolved_form
+                    if unresolved_handle in self.step_dict and unresolved_handle in [
+                        StepHandle.parse_from_key(key) for key in step_keys_to_execute
+                    ]:
+                        # Ok if the entire dynamic step is selected to execute.
+                        # Note: the assumption here is when the entire dynamic step is selected,
+                        # the step_keys_to_execute will include both unresolved step (i.e. [?])
+                        # and all the resolved steps (i.e. [0], ... [n]). Given that at this point
+                        # we no longer track the parent known state (we don't know what "n" was),
+                        # solely from the resolved handles, we can't tell if an entire dynamic
+                        # node is being selected, so the best bet here is to check both unresolved
+                        # and resolved handles exist.
+                        step_handles_to_execute_set.add(unresolved_handle)
+                        continue
                 bad_keys.append(handle.to_key())
 
+            step_handles_to_execute_set.add(handle)
```

Comment:
nit: ^ hard to discern when the bad_keys append fires, maybe a comment

Reply:
added examples to explain when bad keys happen